### DISCUSSION TITLE
Axisymmetry on New Lagrangian Kernels

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#28870] [Q&A Modules: Solid mechanics] [answered]
I saw this: https://github.com/idaholab/moose/issues/20712

Where it seems to have been implemented. But I still got the error:

```
*** ERROR ***
/home/richmond98/projects/purple/Cornea_Linear_Elastic/Cornea_Harmonic_Perturbation_2D_Axisymmetric.i:52.3:
The following error occurred in the Kernel 'div_sig_x_real' of type TotalLagrangianStressDivergence.

This kernel should only act in Cartesian coordinates.
```

I updated MOOSE a few weeks ago. I am using `ComputeLagrangianStrainAxisymmetricCylindrical` for the strain.

**Comment by GiudGiud** [comment]
Hello

Are you creating this object? `'div_sig_x_real' of type TotalLagrangianStressDivergence.`

if not then it s being added by the action. 
Unfortunately this means the SolidMechanics physics is not supporting RZ at the moment. You will need to use the DumpObjectsProblem on a XY problem to get the list of the objects the Physics is creating, then adapt this list manually for RZ coordinates.

@hugary1995 for a double check on whether the new system actually should support RZ / if trying to hack it is a bad idea

**Reply by richmondodufisan** [reply]
Yes I created that object, I didn't use the action set up. This is the block:

```
[Kernels]
  [div_sig_x_real]
    type = TotalLagrangianStressDivergence
	component = 0
	displacements = 'disp_x_real disp_y_real disp_z_real'
    variable = disp_x_real
	base_name = real
  []

...

[]
```

and in my Materials block, I have: (not sure if it is worth checking out but the base_name functionality wasn't implemented for "getting" the elasticity tensor so I had to hardcode it here)

```
[Materials]
  [elastic_tensor_real]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = ${youngs_modulus_val}
    poissons_ratio = ${poissons_ratio_val}
	base_name = real
  []
  [compute_stress_real]
    type = ComputeLagrangianLinearElasticStress
	elasticity_tensor = real_elasticity_tensor
	base_name = real
  []
  [compute_strain_real]
    type = ComputeLagrangianStrainAxisymmetricCylindrical
    displacements = 'disp_x_real disp_y_real disp_z_real'
	base_name = real
  []

...
[]
```

and my mesh block:

```
[Mesh]
  second_order = true
  coord_type = RZ
  [sample_mesh]
    type = FileMeshGenerator
    file = eyeball_2D_axisymmetric.msh
  []
[]

```

**Reply by GiudGiud** [reply]
Then that object that you created does not handle RZ. 

If you can find a substitute that does RZ then try to use that. If not then you would need to implement support for RZ in that object. 

It's not usually very difficult to handle RZ

**Reply by richmondodufisan** [reply, marked as answer]
Hi, I found the kernel (thanks to MOOSE being open source!). It is `TotalLagrangianStressDivergenceAxisymmetricCylindrical`. I think it would be helpful if the documentation page of `TotalLagrangianStressDivergence` links to/mention them (also for the spherical case, `TotalLagrangianStressDivergenceCentrosymmetricSpherical`).

But an important thing I just found (maybe it could be included as a warning?) is that in the axisymmetric case, the 2D mesh MUST be created in the x-y domain. I struggled for a while trying to figure out why the 3D and 2D axisymmetric simulations I ran didn't match, but then tried on a whim to create the mesh in the x-y plane (instead of x-z plane as before) and that fixed it. 

Well, I guess there already is some kind of warning, the page on coordinate systems says that the arrangement is r, z, theta, instead of r, theta, z. But I didn't think it was that obvious till I fixed it.

**Reply by GiudGiud** [reply]
thanks for reporting. I ll create an issue and we ll improve those docs

**Reply by hugary1995** [reply]
Right, thanks to both of you @richmondodufisan @GiudGiud . I agree both the documentation and the action should be improved.
